core/src/main/scala/cats/collections/Predicate.scala

```diff
@@ -46,28 +60,62 @@ abstract class Predicate[-A] extends scala.Function1[A, Boolean] { self =>
   /**
    * Return the opposite predicate
    */
-  def negate: Predicate[A] = Predicate(a => !apply(a))
+  def negate: Predicate[A] = Predicate.wrap[A](p.andThen {
+    case e: Now[Boolean] => if (e.value) Eval.False else Eval.True
+    case e => e.flatMap { if (_) Eval.False else Eval.True }
+  })
```

Comment:
if we had case classes you could defer the code here:
```scala
case class Negate[A](of: Predicate[A]) extends Predicate[A] {
 ...
}
```
then `def negate` could apply the law `negate(negate(p)) == p`. Similar with keeping `And`, `Or` and `Const` nodes around.

Reply:
I'm not really in favor of adding an AST layer.

I feel like most of the time these kinds of optimisations will be spotted by user code that also knows these laws, and that most of the nodes will be opaque predicates (`A => Boolean`) that we cant really do a lot with besides double negation. So the laws would rarely be applied in practice. This also clashes with the type alias route, so simple predicates get an additional indirection, and adds a bunch of boilerplate and complexity.

Side note: would we apply De Morgan's laws in either direction?